libbeat/esleg/eslegclient/connection.go

```diff
@@ -279,17 +275,22 @@ func NewConnectedClient(cfg *cfg.C, beatname string) (*Connection, error) {
 
 // Connect connects the client. It runs a GET request against the root URL of
 // the configured host, updates the known Elasticsearch version and calls
-// globally configured handlers.
-func (conn *Connection) Connect() error {
+// globally configured handlers. The context is used to control the lifecycle
+// of the HTTP requests/connections, the caller is responsible for cancelling
+// the context to stop any in-flight requests.
+func (conn *Connection) Connect(ctx context.Context) error {
 	if conn.log == nil {
 		conn.log = logp.NewLogger("esclientleg")
 	}
+
+	conn.reqsContext = ctx
+
 	if err := conn.getVersion(); err != nil {
 		return err
 	}
 
 	if conn.OnConnectCallback != nil {
-		if err := conn.OnConnectCallback(); err != nil {
+		if err := conn.OnConnectCallback(*conn); err != nil {
```

Comment:
Why can't you pass a reference to `conn`? This isn't actually a deep copy, the reference types in conn (the `maps` for example) are still just having the pointer value copied and point to the same object as before so they could be mutated if that is what you are trying to prevent.

Reply:
No particular reason, I believe it started like that because of my very first implementation was mutating the connection where, then I changed to not mutating it, probably I wanted to ensure this isolation.

I'll update it to receive a pointer.